File: docs-examples/example-java/src/test/java/io/micronaut/nats/docs/consumer/custom/type/ProductInfoSpec.java
package io.micronaut.nats.docs.consumer.custom.type;

import io.micronaut.context.annotation.Property;
import io.micronaut.nats.testcontainers.Nats;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "ProductInfoSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ProductInfoSpec implements TestPropertyProvider {

    @Test
    void testUsingACustomTypeBinder(ProductClient productClient, ProductListener productListener) {

// tag::producer[]
        productClient.send("body".getBytes());
        productClient.send("medium", 20L, "body2".getBytes());
        productClient.send(null, 30L, "body3".getBytes());
// end::producer[]


        await().atMost(60, SECONDS).until(() ->
                productListener.messages.size() == 3 &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 10L) &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 20L) &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 30L)
        );
    }

    @Override
    public Map<String, String> getProperties() {
        return Nats.getProperties();
    }
}

File: docs-examples/example-python/src/test/python/micronaut/nats/docs/consumer/custom/type/ProductInfoSpec.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@Property(name="spec.name", value="ProductInfoSpec")
@MicronautTest(environments=["nats"])
class ProductInfoSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_using_a_custom_type_binder(self):
        # tag::producer[]
        self.product_client.send("body".encode())
        self.product_client.send_with_headers("medium", 20, "body2".encode())
        self.product_client.send_with_headers(None, 30, "body3".encode())
        # end::producer[]

        for _ in range(600):
            if len(self.product_listener.messages) == 3:
                break
            sleep(0.1)
        counts = sorted(int(pi.count) for pi in self.product_listener.messages)
        assert counts == [10, 20, 30]
